### PR TITLE
[Test] Refactor test/test_nn.py to be device-agnostic [8/N]

### PR DESCRIPTION
**Pull request by can-gaa-hou**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #189537

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -15925,6 +15744,176 @@ def test_RNN_cpu_vs_device_with_dropout(self, device):
...
+                if not train:
+                    self.assertEqual(output1.data, prev_output)
+                    self.assertEqual(output2.data, prev_output)
+                else:
+                    self.assertNotEqual(output1.data, prev_output)
+                    self.assertNotEqual(output2.data, prev_output)
+            prev_output = output1.data
+
+    @skipCUDAIfNoCudnn
+    def test_batchnorm_nhwc_train(self, device):
```
Comment (fffrog) [inline test/test_nn.py]: Why do we rename to `test_batchnorm_cudnn_nhwc` from `test_batchnorm_cudnn_nhwc` here?

File: test/test_nn.py [hunk 2]
```diff
@@ -15925,6 +15744,176 @@ def test_RNN_cpu_vs_device_with_dropout(self, device):
...
+        # see #42588, grad is channels_last contiguous, but grad.suggest_memory_format (rightly) return "contiguous"
+        # not channels_last
+        input = torch.randint(1, 10, (2, 8, 8, 1), dtype=torch.float32, device=device)
+        input = input.contiguous(memory_format=torch.channels_last).detach().requires_grad_()
+        grad = torch.randint(1, 10, (2, 8, 8, 1), dtype=torch.float32, device=device)
+        grad = grad.permute(0, 2, 1, 3)
+        run_test(input, grad)
+
+    @skipCUDAIfNoCudnn
+    def test_batchnorm_half(self, device):
```
Comment (fffrog) [inline test/test_nn.py]: ditto

Comment (fffrog) [reply test/test_nn.py]: Got it, you splited the old one into two testcases, which sounds good to me.

File: test/test_nn.py (hunk 1 above)
Comment (can-gaa-hou) [reply test/test_nn.py]: Because there is already a `test_batchnorm_nhwc` testing the "eval" branch, and this refactoring test aims to test the "train" branch. Feel free to suggest a name for this :)

Comment (fffrog) [reply test/test_nn.py]: Thank you for your explanation.

The code logic and semantics are the same as before the change, since we only changed the name, which is no big deal to me.

**Review by fffrog — APPROVED** [review]
SGTM

File: test/test_nn.py [hunk 3]
```diff
@@ -6169,120 +5974,6 @@ def b(N, D, V, prob_target=False, device=cuda):
...
-            mean_ref = torch.mean(data, mean_axes, keepdim=False)
-
-            self.assertEqual(mean_ref, mean1)
-            self.assertEqual(mean_ref, mean2)
-
-        _batch_norm_stats(torch.randn(1, 96, 112, 112, dtype=torch.float, device='cuda'), torch.channels_last, (0, 2, 3))
-        _batch_norm_stats(torch.randn(1, 96, 112, 112, 112, dtype=torch.float, device='cuda'), torch.channels_last_3d, (0, 2, 3, 4))
-
-    @unittest.skipIf(not TEST_CUDA, "CUDA not available")
-    def test_batch_norm_gather_stats_invalid_shapes_cuda(self):
```
Comment (albanD) [inline test/test_nn.py:6253]: This one is deleted?!

File: test/test_nn.py [hunk 4]
```diff
@@ -16079,6 +15870,176 @@ def test_RNN_cpu_vs_device_with_dropout(self, device):
...
+
+        self.assertEqual(hy[0].data.min(), hy[0].data.max())
+        self.assertEqual(hy[1].data.min(), hy[1].data.max())
+        self.assertEqual(hy.data[0][0][0], 10)
+        self.assertEqual(hy.data[1][0][0], output_val)
+
+    @set_default_dtype(torch.double)
+    @parametrize_test('bidirectional', [True, False])
+    @parametrize_test('mode', ['RNN', 'LSTM', 'GRU'])
+    def test_error_RNN_seq_len_zero(self, device, mode, bidirectional):
```
Comment (albanD) [inline test/test_nn.py:16247]: The content of this test is significantly different from the original?

**Review by albanD — CHANGES_REQUESTED** [review]

File: test/test_nn.py (hunk 3 above)
Comment (can-gaa-hou) [reply test/test_nn.py:6253]: Oops, good catch! Add this test to `TestNNDeviceType`. Thanks for the review :)

File: test/test_nn.py (hunk 4 above)
Comment (can-gaa-hou) [reply test/test_nn.py:16247]: Yes, using `parametrize_test` instead of the two for loop, but the logic and coverage should be the same before and after.

**Comment by can-gaa-hou** [comment]
@claude review this pr with pr review skill

**Comment by can-gaa-hou** [comment]
> 4. CPU RNN-dropout coverage dropped on cuDNN machines

this makes sense so I just copy and paste the original code now for cuda test

File: test/test_nn.py [hunk 5]
```diff
@@ -16351,6 +16171,200 @@ def test_RNN_cpu_vs_device_with_dropout(self, device):
         # Because of dropout randomness, can only compare dropout=0 and dropout=1
         self._test_RNN_cpu_vs_device(device, 1)
 
+    @skipCUDAIfNoCudnn
+    @set_default_dtype(torch.double)
+    @parametrize_test('train', [True, False])
+    @parametrize_test('p', [0, 0.276, 0.731, 1])
+    def test_RNN_dropout(self, p, train):
```
Comment (fffrog) [inline test/test_nn.py]: Missing `device` parameter, here?

File: test/test_nn.py [hunk 6]
```diff
@@ -16351,6 +16171,200 @@ def test_RNN_cpu_vs_device_with_dropout(self, device):
...
+            self.assertEqual(hy[0].data.min(), hy[0].data.max())
+            self.assertEqual(hy[1].data.min(), hy[1].data.max())
+            self.assertEqual(hy.data[0][0][0], 10)
+            self.assertEqual(hy.data[1][0][0], output_val)
+
+    @skipCUDAIfNoCudnn
+    @set_default_dtype(torch.double)
+    @parametrize_test('bidirectional', [True, False])
+    @parametrize_test('mode', ['RNN', 'LSTM', 'GRU'])
+    def test_error_RNN_seq_len_zero(self, mode, bidirectional):
```
Comment (fffrog) [inline test/test_nn.py]: ditto

File: test/test_nn.py [hunk 7]
```diff
@@ -16351,6 +16171,200 @@ def test_RNN_cpu_vs_device_with_dropout(self, device):
...
+                rnn.cuda()
+                input = input.cuda()
+
+            with self.assertRaisesRegex(RuntimeError, "Expected sequence length to be larger than 0 in RNN"):
+                rnn(input)
+
+    @skipCUDAIfNoCudnn
+    @parametrize_test('train', [True, False])
+    @parametrize_test('p', [0, 0.1234])
+    def test_RNN_dropout_state(self, p, train):
```
Comment (fffrog) [inline test/test_nn.py]: ditto

File: test/test_nn.py [hunk 8]
```diff
@@ -16351,6 +16171,200 @@ def test_RNN_cpu_vs_device_with_dropout(self, device):
...
+            else:
+                self.assertNotEqual(output1, output2)
+                self.assertNotEqual(output1, output3)
+                self.assertNotEqual(hy1, hy2)
+                self.assertNotEqual(hy1, hy3)
+
+    @skipCUDAIfNoCudnn
+    @set_default_dtype(torch.double)
+    @parametrize_test('train', [True, False])
+    def test_RNN_change_dropout(self, train):
```
Comment (fffrog) [inline test/test_nn.py]: ditto

File: test/test_nn.py [hunk 9]
```diff
@@ -16002,6 +15693,135 @@ def perm_fn(x):
...
+            self.assertEqual(mean_ref, mean1)
+            self.assertEqual(mean_ref, mean2)
+
+        _batch_norm_stats(torch.randn(1, 96, 112, 112, dtype=torch.float, device=device), torch.channels_last, (0, 2, 3))
+        _batch_norm_stats(
+            torch.randn(1, 96, 112, 112, 112, dtype=torch.float, device=device), torch.channels_last_3d, (0, 2, 3, 4))
+
+    @skipMPS
+    @onlyAccelerator
+    def test_batch_norm_gather_stats_invalid_shapes(self, device):
```
Comment (fffrog) [inline test/test_nn.py]: As the suggestion from claude, this testcase should be cuda-specific

File: test/test_nn.py [hunk 10]
```diff
@@ -16351,6 +16171,200 @@ def test_RNN_cpu_vs_device_with_dropout(self, device):
...
+            self.assertEqual(hy.data[0][0][0], 10)
+            self.assertEqual(hy.data[1][0][0], output_val)
+
+    @skipCUDAIfNoCudnn
+    @set_default_dtype(torch.double)
+    @parametrize_test('bidirectional', [True, False])
+    @parametrize_test('mode', ['RNN', 'LSTM', 'GRU'])
+    def test_error_RNN_seq_len_zero(self, mode, bidirectional):
+        # checking error message when RNN has seq_len = 0
+        for device in get_all_device_types():
```
Comment (fffrog) [inline test/test_nn.py]: We can remove this line when we have `device` parameter.

**Review by fffrog — CHANGES_REQUESTED** [review]
Please note the suggetion 4 from claude.

File: test/test_nn.py (hunk 10 above)
Comment (can-gaa-hou) [reply test/test_nn.py]: I think we can't because device here can be cpu and cuda, but if we remove it, it will be cuda only since the test is instantiated only for cuda

File: test/test_nn.py (hunk 9 above)
Comment (can-gaa-hou) [reply test/test_nn.py]: sure fixed

**Comment by can-gaa-hou** [comment]
@claude review this pr with pr review skill

**Comment by fffrog** [comment]
Hi @can-gaa-hou, please follow the suggestions from claude, focus on the issue 1 and issue 2.

**Review by fffrog — DISMISSED** [review]
Thank you.

File: test/test_nn.py [hunk 11]
```diff
@@ -3061,179 +3061,6 @@ def test_projections_errors_on_gru_and_rnn(self):
             with self.assertRaisesRegex(ValueError, error_msg):
                 rnn = getattr(nn, mode)(30, 20, 2, proj_size=10)
 
-    @unittest.skipIf(not TEST_CUDNN, "needs cudnn")
-    @set_default_dtype(torch.double)
-    def test_RNN_dropout(self):
```
Comment (fffrog) [inline test/test_nn.py:3080]: Hi @can-gaa-hou

I think we can keep this testcase as `ACCELERATOR`, which can be runned in CPU and other accelerator including cuda with cudnn, this testcase focus on RNN with dropout(inverted dropout).

As for the decorator, I think we can `skipIf(not Test_CUDNN, "needs cudnn", device="cuda") or `skipCUDAIfNoCudnn`.

File: test/test_nn.py [hunk 12]
```diff
@@ -16445,6 +16201,245 @@ def test_RNN_cpu_vs_device_with_dropout(self, device):
...
+            self.assertEqual(hy[0].data.min(), hy[0].data.max())
+            self.assertEqual(hy[1].data.min(), hy[1].data.max())
+            self.assertEqual(hy.data[0][0][0], 10)
+            self.assertEqual(hy.data[1][0][0], output_val)
+
+    @skipCUDAIfNoCudnn
+    @set_default_dtype(torch.double)
+    @parametrize_test('bidirectional', [True, False])
+    @parametrize_test('mode', ['RNN', 'LSTM', 'GRU'])
+    def test_error_RNN_seq_len_zero(self, device, mode, bidirectional):
```
Comment (fffrog) [inline test/test_nn.py]: I think this one should be `ACCELERATOR` as well.

And we do not need to loop the device in the testcase body, we can just use the `device` parameter to  create the tensor, then do the test logic.

File: test/test_nn.py [hunk 13]
```diff
@@ -3061,179 +3061,6 @@ def test_projections_errors_on_gru_and_rnn(self):
...
-                    rnn = module(5, 6, bidirectional=bidirectional)
-                    if device == 'cuda':
-                        rnn.cuda()
-                        input = input.cuda()
-
-                    with self.assertRaisesRegex(RuntimeError, "Expected sequence length to be larger than 0 in RNN"):
-                        rnn(input)
-
-    @unittest.skipIf(not TEST_CUDNN, "needs cudnn")
-    def test_RNN_dropout_state(self):
```
Comment (fffrog) [inline test/test_nn.py:3139]: ditto

**Review by fffrog — CHANGES_REQUESTED** [review]

File: test/test_nn.py [hunk 14]
```diff
@@ -3075,179 +3075,6 @@ def test_projections_errors_on_gru_and_rnn(self):
...
-                        self.assertEqual(hy1, hy3)
-                    else:
-                        self.assertNotEqual(output1, output2)
-                        self.assertNotEqual(output1, output3)
-                        self.assertNotEqual(hy1, hy2)
-                        self.assertNotEqual(hy1, hy3)
-
-    @unittest.skipIf(not TEST_CUDNN, "needs cudnn")
-    @set_default_dtype(torch.double)
-    def test_RNN_change_dropout(self):
```
Comment (fffrog) [inline test/test_nn.py:3181]: `Accelerator` here?

File: test/test_nn.py [hunk 15]
```diff
@@ -3075,179 +3075,6 @@ def test_projections_errors_on_gru_and_rnn(self):
...
-                    if not train:
-                        self.assertEqual(output1.data, prev_output)
-                        self.assertEqual(output2.data, prev_output)
-                    else:
-                        self.assertNotEqual(output1.data, prev_output)
-                        self.assertNotEqual(output2.data, prev_output)
-                prev_output = output1.data
-
-    @unittest.skipIf(not TEST_CUDA, "CUDA not available")
-    def test_RNN_dropout_gradients(self):
```
Comment (fffrog) [inline test/test_nn.py:3217]: Accelerator here?

File: test/test_nn.py [hunk 16]
```diff
@@ -3802,63 +3629,6 @@ def helper(self, dtype):
...
-        # see #42588, grad is channels_last contiguous, but grad.suggest_memory_format (rightly) return "contiguous"
-        # not channels_last
-        input = torch.randint(1, 10, (2, 8, 8, 1), dtype=torch.float32, device="cuda")
-        input = input.contiguous(memory_format=torch.channels_last).detach().requires_grad_()
-        grad = torch.randint(1, 10, (2, 8, 8, 1), dtype=torch.float32, device="cuda")
-        grad = grad.permute(0, 2, 1, 3)
-        run_test(input, grad)
-
-    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
-    def test_batchnorm_cudnn_half(self):
```
Comment (fffrog) [inline test/test_nn.py:3843]: How about splitting this testcases into two parts:

- `Accelerator`
```Python
input = torch.randint(1, 10, (2, 3, 2, 2), dtype=torch.half, device="cuda", requires_grad=True)
        m = nn.BatchNorm2d(3).half().cuda()
        thnn_output = m(input)
        thnn_output.sum().backward()
        thnn_input_grad = input.grad.data.clone()
        self.assertEqualTypeString(thnn_output, input)
```
- `CUDA` with `CUDNN`
```Python
input.grad = None
            m = m.float()
            cudnn_output = m(input)
            cudnn_output.sum().backward()
            cudnn_input_grad = input.grad.data.clone()
            self.assertEqualTypeString(cudnn_output, input)
            self.assertEqual(cudnn_output, thnn_output)
            self.assertEqual(cudnn_input_grad, thnn_input_grad, atol=1e-3, rtol=0)
```

File: test/test_nn.py [hunk 17]
```diff
@@ -6217,120 +5987,6 @@ def b(N, D, V, prob_target=False, device=cuda):
...
-            mean_ref = torch.mean(data, mean_axes, keepdim=False)
-
-            self.assertEqual(mean_ref, mean1)
-            self.assertEqual(mean_ref, mean2)
-
-        _batch_norm_stats(torch.randn(1, 96, 112, 112, dtype=torch.float, device='cuda'), torch.channels_last, (0, 2, 3))
-        _batch_norm_stats(torch.randn(1, 96, 112, 112, 112, dtype=torch.float, device='cuda'), torch.channels_last_3d, (0, 2, 3, 4))
-
-    @unittest.skipIf(not TEST_CUDA, "CUDA not available")
-    def test_batch_norm_gather_stats_invalid_shapes_cuda(self):
```
Comment (fffrog) [inline test/test_nn.py:6307]: `Accelerator` here?

**Review by fffrog — CHANGES_REQUESTED** [review]

File: test/test_nn.py (hunk 16 above)
Comment (can-gaa-hou) [reply test/test_nn.py:3843]: already split into `test_batchnorm_cudnn_half ` and `test_batchnorm_half `

**Review by fffrog — APPROVED** [review]
SGTM.

All testcases have been checked and no something missing.
